earth2mip/initial_conditions/__init__.py

```diff
@@ -47,13 +55,30 @@ def get_data_source(
         raise NotImplementedError(initial_condition_source)
 
 
-def get_initial_condition_for_model(
-    time_loop: time_loop.TimeLoop, data_source: base.DataSource, time: datetime
+def get_data_from_source(
+    data_source: base.DataSource,
+    time: datetime.datetime,
+    channel_names: List[str],
+    grid: schema.Grid,
+    time_levels: int,
+    time_step: datetime.timedelta = datetime.timedelta(hours=0),
+    device: torch.device = "cpu",
+    dtype: torch.dtype = torch.float,
 ) -> torch.Tensor:
+    """Get data from a data source
+
+    Select the ``channel_names`` and ``regrid`` it to the target ``grid``
 
-    dt = time_loop.history_time_step
+    Args:
+        time_levels: the number of time levels to return
+        time_step: the time step of the time levels
+
+    Returns:
+        (time_levels, c, lat, lon) shaped data
+    """
+    dt = time_step
```

Comment:
It's a worthwhile discussion about data flow, but I think out of scope.  This function is intended to be a private helper that generalizes the `get_initial_condition_for_model`, and is more flexible than that function, and it is needed for graphcast scoring and fixes #156.

Your questions mostly touch more on the general design philosphy of e2mip, and are a bit out of scope for the PR, but in the interest of knowledge transfer I hope these answers are useful:

1. The `.grid` is no longer a schema.Grid Enum. It is an [earth2mip.grid.LatLongGrid](https://github.com/NVIDIA/earth2mip/blob/ebd86637ec853c11e9ff5e717b053351c740e526/earth2mip/grid.py#L25) object, which supports arbitrary lat/lon grids. This was generalization was completed a couple months ago.
2. Time level is roughly described [here](https://github.com/NVIDIA/earth2mip/blob/ebd86637ec853c11e9ff5e717b053351c740e526/earth2mip/time_loop.py#L81), although there is some inconsistent naming.
3. `device`.  Agreed, this and `dtype` probably don't need to be part of this class.

More generally, when fixing bugs are improving performance, I prefer to leave existing APIs/names intact...even if those APIs have some issues...and then fix those APIs in a separate PR.

Reply:
It is not clear that this function is intended to be a private helper. The description and arguments are very general and if I was a user I would not have that context and I might assume that this is intended to be a public function. 

If it is intended to be private then perhaps we should have guidelines for denoting such functions (marking prefix with `_` and clearly denoting it in the function description). 

I think there are two options either we address this in the present PR and clearly mark all of the intended private functions or we make an issue out of it. I will leave the option to you.

For the other functions I agree that we can have a broader discussion about the design of data movement in another forum but sometimes I like having thoughts written for posterity.

Reply:
> If it is intended to be private then perhaps we should have guidelines for denoting such functions 

Agreed, I use the module-level `__all__` for this.

To address your concerns, I have removed this function from that list and added a note to the docstring. I have also renamed the `time_levels` to be more consistent with the `earth2mip.time_loop.TimeLoop` protocol.

device and dtype are still required since it may be beneficial to run the regridding on device.

Reply:
sounds good.